# Tugas P2 Latihan Indexing (Python)

- Mata Kuliah: Sistem Basis Data
- Nama: Ghaisan Khoirul Badruzaman
- NIM: 251524048
- Kelas: 2B-D4

# Intro

Indexing pada Python bertujuan untuk melatih dan memberikan wawasan kepada mahasiswa agar bisa memahami bagaimana data disimpan, diakses, dan dipercepat proses pencariannya melalui teknik indexing.

Index merupakan struktur data tambahan yang digunakan untuk mempercepat proses pencarian pada kumpulan data yang besar. Dengan adanya index, sistem tidak perlu membaca seluruh data (full scan), melainkan cukup menggunakan penunjuk (pointer) ke lokasi data yang relevan.

Struktur indexing yang akan dipelajari mencakup:
1. Dense Index
2. Sparse Index
3. Multi-Level Index dengan Tree (B-Tree)

Praktikum ini menggunakan dataset sample berupa data 200 mahasiswa yang terdiri dari NIM, Nama, dan Nilai.

*Pengayaan (Opsional, nilai tambahan): *

Jika sudah menyelesaikan seluruh bagian praktikum dan ingin eksplorasi lebih lanjut, silahkan coba analisis dampak dari pencarian tanpa indexing dan dengan indexing yang berbeda-beda. Pengayaan juga dapat dilakukan dengan menambah jumlah data, atau implementasi indexing lain seperti B+Tree.



#Import Module

Module Drive digunakan untuk memuat dan meminta izin mengakses direktori dan file pada Google Drive.

Module pandas digunakan untuk memproses data dalam CSV menjadi DataFrame

In [1]:
# Dapat ditambahkan modul lain jika diperlukan

import pandas as pd
# Dijalankan lokal di Jupyter, bukan Colab, jadi Google Drive tidak perlu di-mount.
# Kalau dijalankan di Colab, buka komentar dua baris di bawah.
# from google.colab import drive
# drive.mount('/content/drive')

# Menentukan Path Direktori
Mempersiapkan path direktori yang berisikan kumpulan file CSV pada variable `csv_path`

In [2]:
# Sesuaikan path direktori file CSV masing-masing
# notebook ini dijalankan lokal, mahasiswa.csv ditaruh satu folder dengan notebook
csv_path = '.'
# kalau dijalankan di Colab, buka komentar baris di bawah
# csv_path = '/content/drive/MyDrive/'

#Membuka File CSV

In [3]:
data = pd.read_csv(f"{csv_path}/mahasiswa.csv")

print(data)

      NIM          Nama  Nilai
0    A001    Mahasiswa1     77
1    A002    Mahasiswa2     85
2    A003    Mahasiswa3     69
3    A004    Mahasiswa4     92
4    A005    Mahasiswa5     74
..    ...           ...    ...
195  A196  Mahasiswa196     94
196  A197  Mahasiswa197     67
197  A198  Mahasiswa198     83
198  A199  Mahasiswa199     75
199  A200  Mahasiswa200     90

[200 rows x 3 columns]


# Simulasi Penyimpanan Records (data) pada Block

In [4]:
#Kelompokan records ke block
block_size = 10
blocks=[]

# Write your code here
# setiap block berisi 10 record berurutan, index baris aslinya ikut terbawa
for awal in range(0, len(data), block_size):
    blocks.append(data.iloc[awal:awal + block_size])

for i, block in enumerate(blocks):
    print(f"Block-{i}:")
    print(block)
    print("---")

Block-0:
    NIM         Nama  Nilai
0  A001   Mahasiswa1     77
1  A002   Mahasiswa2     85
2  A003   Mahasiswa3     69
3  A004   Mahasiswa4     92
4  A005   Mahasiswa5     74
5  A006   Mahasiswa6     81
6  A007   Mahasiswa7     88
7  A008   Mahasiswa8     65
8  A009   Mahasiswa9     90
9  A010  Mahasiswa10     72
---
Block-1:
     NIM         Nama  Nilai
10  A011  Mahasiswa11     83
11  A012  Mahasiswa12     78
12  A013  Mahasiswa13     95
13  A014  Mahasiswa14     66
14  A015  Mahasiswa15     87
15  A016  Mahasiswa16     70
16  A017  Mahasiswa17     91
17  A018  Mahasiswa18     76
18  A019  Mahasiswa19     84
19  A020  Mahasiswa20     68
---
Block-2:
     NIM         Nama  Nilai
20  A021  Mahasiswa21     93
21  A022  Mahasiswa22     71
22  A023  Mahasiswa23     82
23  A024  Mahasiswa24     79
24  A025  Mahasiswa25     89
25  A026  Mahasiswa26     67
26  A027  Mahasiswa27     94
27  A028  Mahasiswa28     73
28  A029  Mahasiswa29     86
29  A030  Mahasiswa30     75
---
Block-3:
     N

# Prosedur Create Dense Index

Buat prosedur untuk membuat dense index untuk seluruh row yang ada di `data` dengan struktur

`key = NIM`

`value = (index block, index row dalam block)`

Lalu, cetak dense index yang dihasilkan.



In [5]:
# Membuat Dense Index: setiap record punya entry index
def create_dense_index(blocks, block_size):
    # Write your code here
    dense_index = {}
    nomor = 0   # urutan record di seluruh data, mulai dari 0
    for block in blocks:
        for nim in block["NIM"]:
            # nomor block dan posisi di dalam block dihitung dari urutan record
            dense_index[nim] = (nomor // block_size, nomor % block_size)
            nomor += 1
    return dense_index


# Implementasi pembuatan dense index untuk data yang sudah disimpan dalam blocks
dense_index = create_dense_index(blocks, block_size)
print(dense_index)

{'A001': (0, 0), 'A002': (0, 1), 'A003': (0, 2), 'A004': (0, 3), 'A005': (0, 4), 'A006': (0, 5), 'A007': (0, 6), 'A008': (0, 7), 'A009': (0, 8), 'A010': (0, 9), 'A011': (1, 0), 'A012': (1, 1), 'A013': (1, 2), 'A014': (1, 3), 'A015': (1, 4), 'A016': (1, 5), 'A017': (1, 6), 'A018': (1, 7), 'A019': (1, 8), 'A020': (1, 9), 'A021': (2, 0), 'A022': (2, 1), 'A023': (2, 2), 'A024': (2, 3), 'A025': (2, 4), 'A026': (2, 5), 'A027': (2, 6), 'A028': (2, 7), 'A029': (2, 8), 'A030': (2, 9), 'A031': (3, 0), 'A032': (3, 1), 'A033': (3, 2), 'A034': (3, 3), 'A035': (3, 4), 'A036': (3, 5), 'A037': (3, 6), 'A038': (3, 7), 'A039': (3, 8), 'A040': (3, 9), 'A041': (4, 0), 'A042': (4, 1), 'A043': (4, 2), 'A044': (4, 3), 'A045': (4, 4), 'A046': (4, 5), 'A047': (4, 6), 'A048': (4, 7), 'A049': (4, 8), 'A050': (4, 9), 'A051': (5, 0), 'A052': (5, 1), 'A053': (5, 2), 'A054': (5, 3), 'A055': (5, 4), 'A056': (5, 5), 'A057': (5, 6), 'A058': (5, 7), 'A059': (5, 8), 'A060': (5, 9), 'A061': (6, 0), 'A062': (6, 1), 'A063':

# Prosedur Pencarian dengan Dense Index

In [6]:
def dense_search(dense_index, nim):
    # Write your code here
    print("Step-1: Mencari NIM di index")
    if nim not in dense_index:
        return f"NIM {nim} tidak ditemukan di index"
    block_id, posisi = dense_index[nim]
    print("Step-2: NIM ditemukan, mengambil data dari block sesuai posisi di index")
    print(f"Hasil pencarian {nim} ditemukan di Block {block_id} posisi ke-{posisi}:")
    # langsung ambil baris dari block yang ditunjuk, tanpa membaca block lain
    return blocks[block_id].iloc[posisi]


# Implementasi pencarian NIM "A095"
print (dense_search(dense_index, 'A095'))

Step-1: Mencari NIM di index
Step-2: NIM ditemukan, mengambil data dari block sesuai posisi di index
Hasil pencarian A095 ditemukan di Block 9 posisi ke-4:
NIM             A095
Nama     Mahasiswa95
Nilai             75
Name: 94, dtype: object


# Prosedur Create Sparse Index

Buat prosedur untuk membuat sparse index yang berisi NIM pertama dari setiap block dengan struktur

`key = NIM`

`value = index block`

Lalu, cetak dense index yang dihasilkan.



In [7]:
# Membuat Sparse Index: setiap data pertama di sebuah block disimpan dalam index
def create_sparse_index(blocks):
    # Write your code here
    sparse_index = {}
    for block_id, block in enumerate(blocks):
        sparse_index[block["NIM"].iloc[0]] = block_id
    return sparse_index

sparse_index = create_sparse_index(blocks)
print(sparse_index)

{'A001': 0, 'A011': 1, 'A021': 2, 'A031': 3, 'A041': 4, 'A051': 5, 'A061': 6, 'A071': 7, 'A081': 8, 'A091': 9, 'A101': 10, 'A111': 11, 'A121': 12, 'A131': 13, 'A141': 14, 'A151': 15, 'A161': 16, 'A171': 17, 'A181': 18, 'A191': 19}


# Prosedur Pencarian dengan Sparse Index

In [8]:
def sparse_search(sparse_index, nim):
    # Write your code here
    print("Step-1: Mencari Block terdekat dengan NIM")
    # cari entry terakhir yang NIM-nya <= NIM yang dicari, data sudah urut berdasarkan NIM
    kandidat = None
    for nim_awal, block_id in sparse_index.items():
        if nim_awal <= nim:
            kandidat = block_id
        else:
            break
    if kandidat is None:
        return f"NIM {nim} tidak ditemukan, lebih kecil dari NIM pertama di index"

    print("Step-2: Mencari NIM dalam block kandidat")
    block = blocks[kandidat]
    for posisi, nim_block in enumerate(block["NIM"]):
        if nim_block == nim:
            return block.iloc[posisi]
    return f"NIM {nim} tidak ditemukan di Block {kandidat}"


print (sparse_search(sparse_index, 'A075'))

Step-1: Mencari Block terdekat dengan NIM
Step-2: Mencari NIM dalam block kandidat
NIM             A075
Nama     Mahasiswa75
Nilai             66
Name: 74, dtype: object


# Create B-Tree Index

Buat Kelas B-Tree yang mencakup struktur dan fungsi berikut.

Lalu, cetak traversal index tree yang dihasilkan dari `blocks`.

Catatan: Template di bawah hanya untuk memberikan clue dan boleh dimodifikasi.

In [9]:
# t = minimal jumlah keys setiap node
# Jumlah maksimal keys dalam 1 node = 2t-1
# Orde/Jumlah children setiap node = 2t
# Key pada tugas ini adalah NIM

class BTreeNode:
    def __init__(self, t, leaf=False):
        self.t = t
        self.keys = []        #(NIM, index)
        self.children = []
        self.leaf = leaf

class BTree:
    def __init__(self, t):
        self.root = BTreeNode(t, leaf=True)
        self.t = t

    # Prosedur insert untuk menambahkan key baru (NIM) dan position (block_id, index dalam block), perlu dicek ditambahkan di node mana
    # Jika node sudah ditentukan dan belum penuh, panggil fungsi _insert_non_full
    # Jika node ditentukan dan penuh, maka perlu buat node baru, panggil fungsi split_child, baru _insert_non_full
    def insert(self, key, position):
        root = self.root
        # Node penuh
        if len(root.keys) == (2 * self.t) - 1:
            # Write your code here
            # root baru dibuat di atas root lama, jadi tinggi tree bertambah satu
            root_baru = BTreeNode(self.t)
            root_baru.children.append(root)
            self.root = root_baru
            self.split_child(root_baru, 0)
            self._insert_non_full(root_baru, (key, position))

        # Node tidak penuh
        else:
            self._insert_non_full(root, (key, position))

    # Insert dalam node dilakukan dengan key terurut
    def _insert_non_full(self, node, k):
        # Write your code here
        i = len(node.keys) - 1
        if node.leaf:
            # geser key yang lebih besar ke kanan sampai ketemu tempat untuk k
            node.keys.append(k)
            while i >= 0 and k[0] < node.keys[i][0]:
                node.keys[i + 1] = node.keys[i]
                i -= 1
            node.keys[i + 1] = k
        else:
            while i >= 0 and k[0] < node.keys[i][0]:
                i -= 1
            i += 1
            # child yang mau dimasuki dipecah dulu kalau sudah penuh
            if len(node.children[i].keys) == (2 * self.t) - 1:
                self.split_child(node, i)
                if k[0] > node.keys[i][0]:
                    i += 1
            self._insert_non_full(node.children[i], k)

    # Prosedur untuk memecah node yang sudah penuh (ada tambahan data baru) serta menautkannya kembali ke tree
    def split_child(self, parent, i):
        # Write your code here
        t = self.t
        penuh = parent.children[i]
        kanan = BTreeNode(t, leaf=penuh.leaf)
        # key tengah naik ke parent, t-1 key sisi kanan pindah ke node baru
        parent.keys.insert(i, penuh.keys[t - 1])
        parent.children.insert(i + 1, kanan)
        kanan.keys = penuh.keys[t:]
        penuh.keys = penuh.keys[:t - 1]
        if not penuh.leaf:
            kanan.children = penuh.children[t:]
            penuh.children = penuh.children[:t]

    # Fungsi untuk melakukan pencarian, k=NIM yang dicari
    def search(self, k, node=None):
        # Write your code here
        if node is None:
            node = self.root
        i = 0
        while i < len(node.keys) and k > node.keys[i][0]:
            i += 1
        if i < len(node.keys) and k == node.keys[i][0]:
            block_id, index = node.keys[i][1]
            return blocks[block_id].loc[index]
        if node.leaf:
            return f"NIM {k} tidak ditemukan di B-Tree"
        return self.search(k, node.children[i])

    # Prosedur untuk menampilkan penelusuran tree, boleh dengan pola urutan traversal apapun
    def traverse(self, node=None, level=0):
        # Write your code here
        # preorder: node dicetak dulu, lalu anak-anaknya dengan indentasi satu tingkat lebih dalam
        if node is None:
            node = self.root
        print("   " * level, node.keys)
        for anak in node.children:
            self.traverse(anak, level + 1)

In [10]:
# Contoh Pembuatan Index dalam B-Tree untuk data pada blocks dengan t=3
btree = BTree(t=3)

# Insert seluruh data di blocks ke btree
# Write your code here
# position = (nomor block, index baris di data), dipakai lagi saat search untuk mengambil record
for block_id, block in enumerate(blocks):
    for index, nim in zip(block.index, block["NIM"]):
        print(f"Insert {nim}")
        btree.insert(nim, (block_id, int(index)))

# Tampilkan traversal btree yang sudah terbentuk
print("\nTraversal B-Tree:")
btree.traverse()

Insert A001
Insert A002
Insert A003
Insert A004
Insert A005
Insert A006
Insert A007
Insert A008
Insert A009
Insert A010
Insert A011
Insert A012
Insert A013
Insert A014
Insert A015
Insert A016
Insert A017
Insert A018
Insert A019
Insert A020
Insert A021
Insert A022
Insert A023
Insert A024
Insert A025
Insert A026
Insert A027
Insert A028
Insert A029
Insert A030
Insert A031
Insert A032
Insert A033
Insert A034
Insert A035
Insert A036
Insert A037
Insert A038
Insert A039
Insert A040
Insert A041
Insert A042
Insert A043
Insert A044
Insert A045
Insert A046
Insert A047
Insert A048
Insert A049
Insert A050
Insert A051
Insert A052
Insert A053
Insert A054
Insert A055
Insert A056
Insert A057
Insert A058
Insert A059
Insert A060
Insert A061
Insert A062
Insert A063
Insert A064
Insert A065
Insert A066
Insert A067
Insert A068
Insert A069
Insert A070
Insert A071
Insert A072
Insert A073
Insert A074
Insert A075
Insert A076
Insert A077
Insert A078
Insert A079
Insert A080
Insert A081
Insert A082
Insert A083
Inse

# Implementasi Pencarian dengan BTree Index

In [11]:
# Coba cari data A010 melalui B-Tree
print(btree.search("A010"))

NIM             A010
Nama     Mahasiswa10
Nilai             72
Name: 9, dtype: object


# Pengayaan: Dampak Pencarian Tanpa Index dan Dengan Index

Mengikuti bagian pengayaan di intro, jumlah perbandingan NIM yang dibutuhkan untuk menemukan satu record dihitung
untuk beberapa cara pencarian. Perhitungan dilakukan untuk seluruh 200 NIM, lalu diulang untuk seluruh NIM pada
data 20.000 mahasiswa buatan untuk melihat pengaruh jumlah data.

`dense_search` di atas memakai dict Python, jadi lookup-nya cukup satu langkah karena dict memakai hashing. Supaya
perbandingannya setara, di sini dense index dan sparse index dicari dengan binary search seperti ordered index di
buku Silberschatz. Sparse index juga dihitung dengan scan linear seperti cara `sparse_search` di atas.

In [12]:
import random

# setiap fungsi mengembalikan jumlah perbandingan NIM sampai record ditemukan

def perbandingan_full_scan(semua_nim, nim):
    # tanpa index: NIM dibaca satu per satu dari record pertama
    jumlah = 0
    for n in semua_nim:
        jumlah += 1
        if n == nim:
            break
    return jumlah

def perbandingan_dense(kunci_dense, nim):
    # binary search pada entry dense index yang terurut
    kiri, kanan, jumlah = 0, len(kunci_dense) - 1, 0
    while kiri <= kanan:
        tengah = (kiri + kanan) // 2
        jumlah += 1
        if kunci_dense[tengah] == nim:
            break
        if kunci_dense[tengah] < nim:
            kiri = tengah + 1
        else:
            kanan = tengah - 1
    return jumlah

def isi_block(blocks, block_id, nim):
    # setelah block kandidat ketemu, record di dalam block dicek satu per satu
    jumlah = 0
    for n in blocks[block_id]["NIM"]:
        jumlah += 1
        if n == nim:
            break
    return jumlah

def perbandingan_sparse_linear(blocks, sparse_index, nim):
    # entry sparse index dicek dari awal sampai ketemu entry yang lebih besar dari nim
    jumlah, kandidat = 0, 0
    for nim_awal, block_id in sparse_index.items():
        jumlah += 1
        if nim_awal <= nim:
            kandidat = block_id
        else:
            break
    return jumlah + isi_block(blocks, kandidat, nim)

def perbandingan_sparse_biner(blocks, sparse_index, kunci_sparse, nim):
    # binary search mencari entry terakhir yang <= nim
    kiri, kanan, jumlah, kandidat = 0, len(kunci_sparse) - 1, 0, 0
    while kiri <= kanan:
        tengah = (kiri + kanan) // 2
        jumlah += 1
        if kunci_sparse[tengah] <= nim:
            kandidat = tengah
            kiri = tengah + 1
        else:
            kanan = tengah - 1
    return jumlah + isi_block(blocks, sparse_index[kunci_sparse[kandidat]], nim)

def perbandingan_btree(btree, nim):
    # key di setiap node yang dilewati dari root sampai key ditemukan
    node, jumlah = btree.root, 0
    while True:
        i = 0
        while i < len(node.keys):
            jumlah += 1
            if nim <= node.keys[i][0]:
                break
            i += 1
        if i < len(node.keys) and nim == node.keys[i][0]:
            return jumlah
        node = node.children[i]

def ukuran_btree(node):
    # jumlah key dan jumlah node di seluruh tree
    key, banyak_node = len(node.keys), 1
    for anak in node.children:
        k, n = ukuran_btree(anak)
        key, banyak_node = key + k, banyak_node + n
    return key, banyak_node

def tinggi_btree(btree):
    node, tinggi = btree.root, 1
    while not node.leaf:
        node = node.children[0]
        tinggi += 1
    return tinggi

def bandingkan(judul, blocks, dense_index, sparse_index, btree, daftar_nim):
    semua_nim = [n for block in blocks for n in block["NIM"]]
    kunci_dense, kunci_sparse = list(dense_index), list(sparse_index)
    hasil = {
        "Full scan (tanpa index)": [perbandingan_full_scan(semua_nim, n) for n in daftar_nim],
        "Dense index, binary search": [perbandingan_dense(kunci_dense, n) for n in daftar_nim],
        "Sparse index, scan linear": [perbandingan_sparse_linear(blocks, sparse_index, n) for n in daftar_nim],
        "Sparse index, binary search": [perbandingan_sparse_biner(blocks, sparse_index, kunci_sparse, n)
                                        for n in daftar_nim],
        "B-Tree (t=3)": [perbandingan_btree(btree, n) for n in daftar_nim],
    }
    print(judul)
    print(f"{'Cara pencarian':<30}{'rata-rata':>10}{'terbanyak':>11}")
    for cara, angka in hasil.items():
        print(f"{cara:<30}{sum(angka) / len(angka):>10.2f}{max(angka):>11}")
    key, banyak_node = ukuran_btree(btree.root)
    print(f"Ukuran index: dense {len(dense_index)} entry, sparse {len(sparse_index)} entry, "
          f"B-Tree {key} key dalam {banyak_node} node dengan tinggi {tinggi_btree(btree)} level")
    print()


# 1. data praktikum, seluruh 200 NIM dicari
bandingkan("Data praktikum (200 record, 20 block):", blocks, dense_index, sparse_index, btree, list(data["NIM"]))

# 2. data buatan 20.000 mahasiswa dengan format kolom yang sama, NIM tetap terurut
random.seed(2)
data_besar = pd.DataFrame({
    "NIM": [f"A{i:05d}" for i in range(1, 20001)],
    "Nama": [f"Mahasiswa{i}" for i in range(1, 20001)],
    "Nilai": [random.randint(55, 100) for _ in range(20000)],
})
blocks_besar = [data_besar.iloc[a:a + block_size] for a in range(0, len(data_besar), block_size)]
dense_besar = create_dense_index(blocks_besar, block_size)
sparse_besar = create_sparse_index(blocks_besar)
btree_besar = BTree(t=3)
for block_id, block in enumerate(blocks_besar):
    for index, nim in zip(block.index, block["NIM"]):
        btree_besar.insert(nim, (block_id, int(index)))

# seluruh 20.000 NIM dicari supaya rata-rata dan kasus terburuknya pasti, bukan perkiraan dari sampel
bandingkan("Data buatan (20.000 record, 2.000 block), seluruh NIM dicari:",
           blocks_besar, dense_besar, sparse_besar, btree_besar, list(data_besar["NIM"]))

Data praktikum (200 record, 20 block):
Cara pencarian                 rata-rata  terbanyak
Full scan (tanpa index)           100.50        200
Dense index, binary search          6.76          8
Sparse index, scan linear          16.95         30
Sparse index, binary search        10.00         15
B-Tree (t=3)                        7.28         14
Ukuran index: dense 200 entry, sparse 20 entry, B-Tree 200 key dalam 98 node dengan tinggi 5 level



Data buatan (20.000 record, 2.000 block), seluruh NIM dicari:
Cara pencarian                 rata-rata  terbanyak
Full scan (tanpa index)         10000.50      20000
Dense index, binary search         13.36         15
Sparse index, scan linear        1007.00       2010
Sparse index, binary search        16.48         21
B-Tree (t=3)                       14.06         26
Ukuran index: dense 20000 entry, sparse 2000 entry, B-Tree 20000 key dalam 9996 node dengan tinggi 9 level



## Analisis Pengayaan

1. Tanpa index, rata-rata NIM yang harus dibandingkan kira-kira setengah dari jumlah data: 100,50 dari 200 record
   dan 10.000,50 dari 20.000 record, dengan kasus terburuk seluruh data dibaca. Jumlah langkahnya naik lurus
   mengikuti jumlah data, jadi data 100 kali lebih banyak berarti pencarian kira-kira 100 kali lebih lama.
2. Dengan index yang dicari secara efisien, jumlah perbandingan hanya tumbuh logaritmik. Saat data naik 100 kali,
   rata-rata dense index dengan binary search naik dari 6,76 ke 13,36, sparse index dengan binary search dari
   10,00 ke 16,48, dan B-Tree dari 7,28 ke 14,06. Kasus terburuknya pun paling banyak 26 perbandingan, sedangkan
   full scan bisa sampai 20.000.
3. Dense index punya satu entry untuk setiap record, jadi jumlah entry-nya sama dengan jumlah record (200 dan
   20.000 entry). Satu entry hanya berisi NIM dan penunjuk posisi, jadi tetap lebih kecil daripada record lengkap.
   `dense_search` di atas cukup satu langkah karena memakai dict Python yang berbasis hashing. Sebagai ordered
   index di buku, dense index dicari dengan binary search sehingga langkahnya sekitar log2 n (paling banyak 8
   untuk 200 record dan 15 untuk 20.000 record).
4. Sparse index hanya berisi NIM pertama tiap block, jadi jumlah entry-nya sepersepuluh jumlah record (20 dan
   2.000 entry) karena satu block berisi 10 record. Perbandingannya sedikit lebih banyak daripada dense index
   karena setelah block kandidat ketemu, isi block masih dicek satu per satu. Cara mencari di index-nya sendiri
   sangat berpengaruh: kalau entry sparse index dicek satu per satu seperti `sparse_search` di atas, rata-ratanya
   naik menjadi 1.007,00 untuk 20.000 record, sedangkan dengan binary search cukup 16,48.
5. Binary search butuh index yang terurut dan sebaiknya ada di memori. Kalau index sudah terlalu besar dan
   tersimpan di disk, binary search masih harus membaca banyak block index. Karena itu index yang besar dibuat
   bertingkat (multi-level index), yaitu index dibuatkan sparse index lagi di atasnya supaya block index yang
   dibaca lebih sedikit. B+Tree adalah bentuk bertingkat seperti ini. B-Tree memakai ide yang sama, yaitu node di
   atas mengarahkan pencarian ke node di bawahnya sehingga tiap level cukup dibaca satu node. Bedanya, key di node
   atas B-Tree tidak diulang di leaf. Tinggi B-Tree di sini hanya naik dari 5 ke 9 level saat data naik 100 kali.
   Karena NIM dimasukkan terurut, sebagian besar node hanya berisi 2 key (batas minimum untuk t=3), jadi tree-nya
   lebih tinggi daripada kalau node terisi penuh.
6. B-Tree menyimpan setiap NIM tepat satu kali (200 key dalam 98 node dan 20.000 key dalam 9.996 node), jadi
   jumlah entry-nya sama dengan dense index. Kelebihannya ada saat data baru dimasukkan. B-Tree cukup menyisipkan
   key ke satu leaf, dan pergeseran key hanya terjadi di dalam node itu (paling banyak 5 key). Kalau leaf penuh,
   node dipecah dan key tengahnya naik ke parent, jadi perubahannya hanya di satu node dan jalur ke root, dan
   semua leaf tetap di level yang sama. Pada file index yang terurut, semua entry setelah posisi sisip harus
   digeser, bisa sampai seluruh isi file. Key di B-Tree juga terurut sehingga rentang NIM bisa dicari lewat
   traversal inorder, walaupun untuk pencarian rentang B+Tree lebih cocok karena semua key ada di leaf yang
   saling terhubung.

Kesimpulannya, index menukar ruang penyimpanan tambahan dengan pencarian yang jauh lebih cepat. Dense index dan
B-Tree punya entry sebanyak record, sparse index punya entry paling sedikit, dan ketiganya membuat pencarian
tumbuh logaritmik selama index-nya sendiri dicari dengan cara yang efisien. B-Tree unggul ketika data sering
bertambah karena tree-nya tetap seimbang tanpa harus menggeser seluruh isi index.

# Pengayaan: Implementasi B+Tree

Intro juga menyarankan implementasi index lain seperti B+Tree. Bedanya dengan B-Tree di atas:

1. Semua NIM beserta position-nya disimpan di leaf. Node internal hanya berisi key pemisah untuk mengarahkan
   pencarian, dan key pemisah itu adalah salinan NIM terkecil di subtree sebelah kanannya (untuk node internal
   tepat di atas leaf, berarti NIM pertama leaf di kanannya).
2. Leaf saling terhubung lewat pointer `next`, jadi pencarian rentang NIM cukup turun sekali ke leaf awal lalu
   berjalan ke kanan tanpa naik turun level.
3. Saat leaf dipecah, key pertama leaf kanan disalin ke parent. Saat node internal dipecah, key tengahnya
   dipindah ke atas seperti di B-Tree.

Supaya bisa dibandingkan dengan B-Tree t=3, jumlah key maksimal di setiap node B+Tree juga dibuat 5.

In [13]:
# B+Tree: node internal hanya berisi key pemisah, semua NIM dan position ada di leaf

class BPlusNode:
    def __init__(self, leaf=False):
        self.leaf = leaf
        self.keys = []       # leaf: NIM, node internal: key pemisah
        self.values = []     # hanya dipakai di leaf: position (block_id, index)
        self.children = []   # hanya dipakai di node internal
        self.next = None     # hanya dipakai di leaf: leaf di sebelah kanannya

class BPlusTree:
    def __init__(self, blocks, max_keys=5):
        self.root = BPlusNode(leaf=True)
        self.blocks = blocks          # sumber record yang ditunjuk position
        self.max_keys = max_keys      # 5 disamakan dengan B-Tree t=3 (2t-1 key)

    # turun dari root ke leaf yang mungkin berisi key, jalurnya disimpan untuk proses split
    def _turun_ke_leaf(self, key):
        node, jalur = self.root, []
        while not node.leaf:
            i = 0
            while i < len(node.keys) and key >= node.keys[i]:
                i += 1
            jalur.append((node, i))
            node = node.children[i]
        return node, jalur

    def insert(self, key, position):
        leaf, jalur = self._turun_ke_leaf(key)
        i = 0
        while i < len(leaf.keys) and leaf.keys[i] < key:
            i += 1
        leaf.keys.insert(i, key)
        leaf.values.insert(i, position)
        if len(leaf.keys) > self.max_keys:
            # leaf dipecah dua, key pertama leaf kanan disalin ke parent sebagai pemisah
            tengah = len(leaf.keys) // 2
            kanan = BPlusNode(leaf=True)
            kanan.keys, kanan.values = leaf.keys[tengah:], leaf.values[tengah:]
            leaf.keys, leaf.values = leaf.keys[:tengah], leaf.values[:tengah]
            kanan.next, leaf.next = leaf.next, kanan
            self._naikkan(jalur, leaf, kanan.keys[0], kanan)

    # memasukkan key pemisah ke parent, parent ikut dipecah kalau kepenuhan
    def _naikkan(self, jalur, kiri, key, kanan):
        if not jalur:
            # yang dipecah adalah root, jadi dibuat root baru dan tinggi tree bertambah satu
            root = BPlusNode()
            root.keys, root.children = [key], [kiri, kanan]
            self.root = root
            return
        parent, i = jalur.pop()
        parent.keys.insert(i, key)
        parent.children.insert(i + 1, kanan)
        if len(parent.keys) > self.max_keys:
            # node internal dipecah, key tengah dipindah ke atas (tidak disalin seperti di leaf)
            tengah = len(parent.keys) // 2
            naik = parent.keys[tengah]
            baru = BPlusNode()
            baru.keys, baru.children = parent.keys[tengah + 1:], parent.children[tengah + 1:]
            parent.keys, parent.children = parent.keys[:tengah], parent.children[:tengah + 1]
            self._naikkan(jalur, parent, naik, baru)

    def search(self, key):
        leaf, _ = self._turun_ke_leaf(key)
        for k, (block_id, index) in zip(leaf.keys, leaf.values):
            if k == key:
                return self.blocks[block_id].loc[index]
        return f"NIM {key} tidak ditemukan di B+Tree"

    # turun sekali ke leaf tempat NIM awal, lalu berjalan ke kanan lewat next sampai melewati NIM akhir
    def range_search(self, awal, akhir):
        leaf, _ = self._turun_ke_leaf(awal)
        posisi = []
        while leaf is not None:
            for k, value in zip(leaf.keys, leaf.values):
                if k > akhir:
                    leaf = None
                    break
                if k >= awal:
                    posisi.append(value)
            else:
                leaf = leaf.next
        if not posisi:
            return self.blocks[0].iloc[0:0]
        return pd.concat([self.blocks[b].loc[[index]] for b, index in posisi])

    def leaf_pertama(self):
        node = self.root
        while not node.leaf:
            node = node.children[0]
        return node


# membangun B+Tree dari seluruh data di blocks, urutannya sama dengan B-Tree
bplus = BPlusTree(blocks, max_keys=5)
for block_id, block in enumerate(blocks):
    for index, nim in zip(block.index, block["NIM"]):
        bplus.insert(nim, (block_id, int(index)))

def tampilkan_internal(node, level=0):
    # hanya node internal yang dicetak, isi leaf dicetak terpisah lewat rantai next
    if node.leaf:
        return
    print("   " * level, node.keys)
    for anak in node.children:
        tampilkan_internal(anak, level + 1)

print("Node internal B+Tree (hanya key pemisah):")
tampilkan_internal(bplus.root)

isi_leaf = []
leaf = bplus.leaf_pertama()
while leaf is not None:
    isi_leaf.append(leaf.keys)
    leaf = leaf.next
print(f"\nJumlah leaf: {len(isi_leaf)}, jumlah NIM di seluruh leaf: {sum(len(k) for k in isi_leaf)}")
print("3 leaf pertama :", isi_leaf[:3])
print("2 leaf terakhir:", isi_leaf[-2:])

Node internal B+Tree (hanya key pemisah):
 ['A049', 'A097', 'A145']
    ['A013', 'A025', 'A037']
       ['A004', 'A007', 'A010']
       ['A016', 'A019', 'A022']
       ['A028', 'A031', 'A034']
       ['A040', 'A043', 'A046']
    ['A061', 'A073', 'A085']
       ['A052', 'A055', 'A058']
       ['A064', 'A067', 'A070']
       ['A076', 'A079', 'A082']
       ['A088', 'A091', 'A094']
    ['A109', 'A121', 'A133']
       ['A100', 'A103', 'A106']
       ['A112', 'A115', 'A118']
       ['A124', 'A127', 'A130']
       ['A136', 'A139', 'A142']
    ['A157', 'A169', 'A181']
       ['A148', 'A151', 'A154']
       ['A160', 'A163', 'A166']
       ['A172', 'A175', 'A178']
       ['A184', 'A187', 'A190', 'A193', 'A196']

Jumlah leaf: 66, jumlah NIM di seluruh leaf: 200
3 leaf pertama : [['A001', 'A002', 'A003'], ['A004', 'A005', 'A006'], ['A007', 'A008', 'A009']]
2 leaf terakhir: [['A193', 'A194', 'A195'], ['A196', 'A197', 'A198', 'A199', 'A200']]


In [14]:
# Pencarian satu NIM dan pencarian rentang lewat B+Tree
print("Pencarian NIM A010 lewat B+Tree:")
print(bplus.search("A010"))
print()
print(bplus.search("A201"))
print()
print("Pencarian rentang NIM A095 sampai A105 lewat leaf yang saling terhubung:")
print(bplus.range_search("A095", "A105"))
print()

# pengecekan kebenaran B+Tree
def kedalaman_leaf(node, level=1):
    if node.leaf:
        return {level}
    hasil = set()
    for anak in node.children:
        hasil |= kedalaman_leaf(anak, level + 1)
    return hasil

rantai = [k for keys in isi_leaf for k in keys]
sama = sum(bplus.search(n).equals(btree.search(n)) for n in data["NIM"])
rentang_uji = [("A001", "A200"), ("A095", "A105"), ("A150", "A150"), ("A000", "A003"), ("A198", "A999"),
               ("A2000", "A2001")]
rentang_benar = all(
    bplus.range_search(a, b).equals(data[(data["NIM"] >= a) & (data["NIM"] <= b)]) for a, b in rentang_uji)
print("Rantai leaf berisi seluruh NIM secara terurut:", rantai == sorted(data["NIM"]))
print("Semua leaf berada di kedalaman yang sama     :", len(kedalaman_leaf(bplus.root)) == 1)
print(f"Record B+Tree sama dengan B-Tree untuk {sama} dari {len(data)} NIM")
print(f"Hasil {len(rentang_uji)} pencarian rentang sama dengan filter langsung pada data:", rentang_benar)

Pencarian NIM A010 lewat B+Tree:
NIM             A010
Nama     Mahasiswa10
Nilai             72
Name: 9, dtype: object

NIM A201 tidak ditemukan di B+Tree

Pencarian rentang NIM A095 sampai A105 lewat leaf yang saling terhubung:
      NIM          Nama  Nilai
94   A095   Mahasiswa95     75
95   A096   Mahasiswa96     93
96   A097   Mahasiswa97     67
97   A098   Mahasiswa98     82
98   A099   Mahasiswa99     79
99   A100  Mahasiswa100     94
100  A101  Mahasiswa101     72
101  A102  Mahasiswa102     85
102  A103  Mahasiswa103     78
103  A104  Mahasiswa104     90
104  A105  Mahasiswa105     66

Rantai leaf berisi seluruh NIM secara terurut: True
Semua leaf berada di kedalaman yang sama     : True
Record B+Tree sama dengan B-Tree untuk 200 dari 200 NIM
Hasil 6 pencarian rentang sama dengan filter langsung pada data: True


In [15]:
# Perbandingan B-Tree dan B+Tree: pencarian satu NIM, pencarian rentang, dan ukuran tree

def perbandingan_bplus(tree, nim):
    # key pemisah yang dibandingkan di setiap node internal, ditambah key di leaf sampai ketemu
    node, jumlah = tree.root, 0
    while not node.leaf:
        i = 0
        while i < len(node.keys):
            jumlah += 1
            if nim < node.keys[i]:
                break
            i += 1
        node = node.children[i]
    for k in node.keys:
        jumlah += 1
        if k == nim:
            break
    return jumlah

def kunjungan_rentang_btree(node, awal, akhir):
    # inorder traversal yang dipangkas, hasilnya (jumlah node dikunjungi, jumlah NIM dalam rentang)
    kunjungan, ketemu = 1, 0
    for i, (k, _) in enumerate(node.keys):
        if not node.leaf and awal < k:
            a, b = kunjungan_rentang_btree(node.children[i], awal, akhir)
            kunjungan, ketemu = kunjungan + a, ketemu + b
        if k > akhir:
            return kunjungan, ketemu
        if k >= awal:
            ketemu += 1
    if not node.leaf:
        a, b = kunjungan_rentang_btree(node.children[-1], awal, akhir)
        kunjungan, ketemu = kunjungan + a, ketemu + b
    return kunjungan, ketemu

def kunjungan_rentang_bplus(tree, awal, akhir):
    # turun sekali ke leaf awal, lalu hanya berjalan ke kanan lewat next
    node, kunjungan, ketemu = tree.root, 0, 0
    while not node.leaf:
        kunjungan += 1
        i = 0
        while i < len(node.keys) and awal >= node.keys[i]:
            i += 1
        node = node.children[i]
    while node is not None:
        kunjungan += 1
        for k in node.keys:
            if k > akhir:
                return kunjungan, ketemu
            if k >= awal:
                ketemu += 1
        node = node.next
    return kunjungan, ketemu

def statistik(node):
    # jumlah node dan jumlah key yang tersimpan, dipakai untuk B-Tree maupun B+Tree
    # (di B+Tree key pemisah di node internal ikut terhitung)
    banyak_node, key = 1, len(node.keys)
    for anak in node.children:
        n, k = statistik(anak)
        banyak_node, key = banyak_node + n, key + k
    return banyak_node, key

def tinggi(node):
    t = 1
    while not node.leaf:
        node = node.children[0]
        t += 1
    return t

def bandingkan_tree(judul, btree_x, bplus_x, semua_nim, daftar_rentang):
    print(judul)
    for nama, tree, fungsi in [("B-Tree", btree_x, perbandingan_btree), ("B+Tree", bplus_x, perbandingan_bplus)]:
        angka = [fungsi(tree, n) for n in semua_nim]
        banyak_node, key = statistik(tree.root)
        print(f"  {nama}: tinggi {tinggi(tree.root)} level, {banyak_node} node, {key} key tersimpan, "
              f"pencarian satu NIM rata-rata {sum(angka) / len(angka):.2f} perbandingan (terbanyak {max(angka)})")
    print(f"  {'Rentang NIM':<22}{'jumlah NIM':>11}{'node B-Tree':>13}{'node B+Tree':>13}")
    for awal, akhir in daftar_rentang:
        node_b, ketemu_b = kunjungan_rentang_btree(btree_x.root, awal, akhir)
        node_p, ketemu_p = kunjungan_rentang_bplus(bplus_x, awal, akhir)
        assert ketemu_b == ketemu_p
        print(f"  {awal + ' - ' + akhir:<22}{ketemu_b:>11}{node_b:>13}{node_p:>13}")
    print()


bandingkan_tree("Data praktikum (200 record):", btree, bplus, list(data["NIM"]),
                [("A095", "A105"), ("A050", "A149"), ("A001", "A200")])

# data buatan 20.000 record dari bagian pengayaan sebelumnya
bplus_besar = BPlusTree(blocks_besar, max_keys=5)
for block_id, block in enumerate(blocks_besar):
    for index, nim in zip(block.index, block["NIM"]):
        bplus_besar.insert(nim, (block_id, int(index)))
rentang_besar = [("A05000", "A05009"), ("A05000", "A05099"), ("A05000", "A05999"), ("A00001", "A20000")]
bandingkan_tree("Data buatan (20.000 record), NIM dimasukkan terurut:", btree_besar, bplus_besar,
                list(data_besar["NIM"]), rentang_besar)

# urutan insert diacak untuk melihat seberapa besar pengaruh aturan pemecahan node saat NIM masuk terurut
urutan_acak = [(block_id, int(index), nim) for block_id, block in enumerate(blocks_besar)
               for index, nim in zip(block.index, block["NIM"])]
random.Random(3).shuffle(urutan_acak)
btree_acak, bplus_acak = BTree(t=3), BPlusTree(blocks_besar, max_keys=5)
for block_id, index, nim in urutan_acak:
    btree_acak.insert(nim, (block_id, index))
    bplus_acak.insert(nim, (block_id, index))
bandingkan_tree("Data buatan (20.000 record), NIM dimasukkan dengan urutan acak:", btree_acak, bplus_acak,
                list(data_besar["NIM"]), rentang_besar)

Data praktikum (200 record):
  B-Tree: tinggi 5 level, 98 node, 200 key tersimpan, pencarian satu NIM rata-rata 7.28 perbandingan (terbanyak 14)
  B+Tree: tinggi 4 level, 87 node, 265 key tersimpan, pencarian satu NIM rata-rata 8.96 perbandingan (terbanyak 16)
  Rentang NIM            jumlah NIM  node B-Tree  node B+Tree
  A095 - A105                    11           10            8
  A050 - A149                   100           54           37
  A001 - A200                   200           98           69

Data buatan (20.000 record), NIM dimasukkan terurut:
  B-Tree: tinggi 9 level, 9996 node, 20000 key tersimpan, pencarian satu NIM rata-rata 14.06 perbandingan (terbanyak 26)
  B+Tree: tinggi 7 level, 8885 node, 26665 key tersimpan, pencarian satu NIM rata-rata 16.98 perbandingan (terbanyak 31)
  Rentang NIM            jumlah NIM  node B-Tree  node B+Tree
  A05000 - A05009                10           13           10
  A05000 - A05099               100           57           40
  A05000 

Data buatan (20.000 record), NIM dimasukkan dengan urutan acak:
  B-Tree: tinggi 8 level, 6223 node, 20000 key tersimpan, pencarian satu NIM rata-rata 15.25 perbandingan (terbanyak 25)


  B+Tree: tinggi 7 level, 7028 node, 25417 key tersimpan, pencarian satu NIM rata-rata 17.29 perbandingan (terbanyak 29)
  Rentang NIM            jumlah NIM  node B-Tree  node B+Tree
  A05000 - A05009                10           10           10
  A05000 - A05099               100           40           35
  A05000 - A05999              1000          320          271
  A00001 - A20000             20000         6223         5424



## Analisis Pengayaan B+Tree

1. Struktur. B+Tree menyimpan 265 key untuk 200 NIM dan 26.665 key untuk 20.000 NIM, karena key pemisah di node
   internal adalah salinan NIM yang juga ada di leaf. Contohnya, A049 di root B+Tree juga muncul di salah satu
   leaf. B-Tree menyimpan setiap NIM tepat sekali (200 dan 20.000 key). Walaupun key-nya lebih banyak, saat NIM
   dimasukkan terurut jumlah node B+Tree justru lebih sedikit (87 lawan 98 dan 8.885 lawan 9.996 node) karena
   node-nya lebih penuh. Dengan urutan insert acak, B+Tree butuh lebih banyak node (7.028 lawan 6.223), jadi key
   pemisah memang menambah kebutuhan ruang.
2. Tinggi tree. Saat NIM dimasukkan terurut, B+Tree lebih pendek (4 lawan 5 level untuk 200 data dan 7 lawan 9
   level untuk 20.000 data). Penyebabnya aturan pemecahan node di kedua implementasi ini, bukan sifat B+Tree itu
   sendiri. B-Tree memecah node yang sudah penuh (5 key) sebelum insert menjadi 2 key, 1 key naik, dan 2 key, lalu
   node kiri tidak pernah menerima NIM baru lagi, jadi sebagian besar node B-Tree tinggal berisi 2 key. B+Tree
   baru memecah leaf saat isinya 6 key menjadi 3 dan 3, jadi leaf dan node internalnya berisi 3 key dan setiap
   node punya lebih banyak anak. Dengan urutan insert acak, bedanya mengecil menjadi 7 lawan 8 level. Di disk,
   node internal B+Tree hanya berisi key tanpa position record sehingga bisa memuat lebih banyak key, tetapi di
   simulasi ini jumlah key maksimal per node sengaja disamakan (5) sehingga kelebihan itu tidak ikut terlihat.
3. Pencarian satu NIM. Rata-rata perbandingan B+Tree sedikit lebih banyak (8,96 lawan 7,28 untuk 200 data,
   16,98 lawan 14,06 untuk 20.000 data terurut, dan 17,29 lawan 15,25 untuk urutan acak). Penyebab utamanya isi
   node: key di dalam node dicek satu per satu dan node B+Tree lebih penuh, jadi di setiap node yang dilewati
   B+Tree membandingkan lebih banyak key. B+Tree selalu membaca node sebanyak tingginya (4 atau 7 node), sedangkan
   B-Tree bisa berhenti di node internal kalau NIM-nya ada di sana, tetapi bisa juga membaca sampai 5 atau 9 node.
4. Pencarian rentang. B+Tree membaca lebih sedikit node di semua percobaan. Pada data 200 record, rentang 100 NIM
   dibaca B+Tree dengan 37 node dan B-Tree dengan 54 node. Pada data 20.000 record yang dimasukkan terurut, rentang
   1.000 NIM butuh 340 lawan 509 node, dan membaca seluruh 20.000 NIM butuh 6.672 lawan 9.996 node. Pada data
   terurut ini hampir seluruh selisih datang dari leaf B+Tree yang lebih padat: satu leaf berisi 3 NIM sedangkan
   satu node B-Tree rata-rata 2 NIM, dan 9.996 dibagi 6.672 kira-kira 1,5 atau 3 dibagi 2. Dengan urutan insert
   acak selisihnya mengecil, misalnya 271 lawan 320 node untuk rentang 1.000 NIM, tetapi B+Tree tetap lebih
   sedikit. Penghitung di sini menghitung setiap node sekali saja, jadi kembalinya traversal inorder B-Tree ke node
   internal tidak ikut terhitung. Perbedaan pola aksesnya tetap ada: B+Tree cukup turun sekali ke leaf awal lalu
   berjalan ke kanan lewat pointer next, sedangkan B-Tree harus bolak-balik antara leaf dan node internal karena
   sebagian NIM tersimpan di node internal.
5. Kebenaran. Rantai leaf berisi 200 NIM secara terurut, semua leaf berada di kedalaman yang sama, record hasil
   B+Tree sama dengan hasil B-Tree untuk seluruh 200 NIM, dan enam pencarian rentang yang diuji memberi hasil yang
   sama dengan filter langsung pada data.

Kesimpulannya, B+Tree menyimpan key tambahan untuk key pemisah, tetapi pencarian rentang dan pembacaan berurutannya
lebih sederhana karena cukup berjalan di rantai leaf. Di percobaan ini B+Tree juga membaca lebih sedikit node untuk
pencarian rentang, tetapi sebagian besar keunggulan angka itu datang dari node B+Tree yang lebih penuh akibat
aturan pemecahan node, terlihat dari selisihnya yang mengecil saat urutan insert diacak. Materi menyebut B+Tree
menyimpan pointer data hanya di leaf dan mayoritas dipakai DBMS untuk indexing. Percobaan ini memperlihatkan salah
satu alasannya, yaitu query rentang seperti mencari NIM dari A050 sampai A149 atau menampilkan data terurut cukup
dilayani oleh rantai leaf.